### PR TITLE
Drop ARP requests based on specific scenarios

### LINKED ISSUE #11329: net: arp: Zephyr replies to multicast address cause of malfored ARP request
Zephyr replies to IPv4 packets from multicast address after reception of ARP Request with Hardware and Protocol Sender Address set to multicast. This ARP Request should be dropped instead and translation table should not be updated.

**Note: Zephyr sets source IP address to 0.0.0.0 in all of the responses (TCP, UDP, ICMP).**

[ipv4_arp_multicast_udp.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/2576324/ipv4_arp_multicast_udp.pcap.gz)
[ipv4_arp_multicast_tcp.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/2576325/ipv4_arp_multicast_tcp.pcap.gz)
[ipv4_arp_multicast_icmp.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/2576326/ipv4_arp_multicast_icmp.pcap.gz)

### LINKED ISSUE #11489: net: arp: Zephyr replies to localhost address
Zephyr replies to ARP Request with sender IP address set to localhost. This ARP Request should be dropped instead and translation table should not be updated

[arp-localhost.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/2594851/arp-localhost.pcap.gz)

### PR DESCRIPTION
**Pull request by rveerama1**
This PR handles special ARP requests. Based on specific scenarios ARP requests will be dropped silently.
Silently discard ARP request if Ethernet address is broadcast and Source IP address is Multicast  adress.
REFERENCES:
    RFC 826: page 4
    RFC 1122: section 3.3.6 {Silently discard link-layer-only b'cast dg's}
    RFC 1812: section 3.3.2, page 34

Networking stack sometimes try to find source address based on destination address. If interface could not find best match then it returns unspecified address (0.0.0.0). Host should not send these packets.

Drop  ARP request with sender IP address set to localhost.

Fixes #11329.
Fixes #11489

### CODE REVIEW

**Review by tbursztyka — DISMISSED** [review]

**Comment by rveerama1** [comment]
One of the address combination in TCP connection still fails, I am working on it.

**Review by pfalcon — CHANGES_REQUESTED** [review]
@rveerama1 : I'm sorry, but I would like to see logging explaining why a packet is being dropped. Just like week I spent an hour or so figuring why my packets are dropped silently. Well, that time it turned out to be a logging problem, but can anticipate it happen again with changes like that.

The commit with Eth broadcast address perhaps that as is, but description like "Networking stack sometimes try to find source address based
on destination address. If interface could not find best match
then it returns unspecified address (0.0.0.0)" really makes me wary, please add a logging there (or explain where/how it already happens). Thanks.

**Comment by rveerama1** [comment]
>  I'm sorry, but I would like to see logging explaining why a packet is being dropped

I have added comments in source, but not logging anything.

**Comment by rveerama1** [comment]
> Just like week I spent an hour or so figuring why my packets are dropped silently.

There are many places in if conditions returns some errors. Not every failure if condition has debug logging.

**Comment by tbursztyka** [comment]
It would be quite painful to log every place where a packet can be dropped, there are legions of reasons for it. Instead I would revisit ethernet recv function so it would return only at the very end of the function, and if NET_DROP is the verdict it would log it. something like that. So nothing for this PR.

@pfalcon if you could dismiss your review on that

File: subsys/net/ip/net_context.c
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: Here's where I suggest to add debug logging. Unless you can point to to the fact that net_if_ipv4_select_src_addr() call above already does *clear* logging (like, it hints that if it fails, then some packets can be not sent out, which I guess it doesn't).

**Comment by pfalcon** [comment]
> It would be quite painful to log every place where a packet can be dropped

That's why I don't call to do that. What I point it is that there's latest fashion to not log anything at all (like this patch does), and ask to log at least *somewhere*, where the probability of hitting it is (speculatively) high. I added inline comment where I think it's worth to add it, feel free to correct me if it doesn't make sense at that place.

**Comment by rveerama1** [comment]
@tbursztyka

**Review by rmstoi — APPROVED** [review]

**Comment by rveerama1** [comment]
@tbursztyka ping

File: subsys/net/ip/tcp.c
Comment (tbursztyka) [inline subsys/net/ip/tcp.c]: How is this related to #11329, as stated in the commit message?

Comment (rveerama1) [reply subsys/net/ip/tcp.c]: ipv4_arp_multicast_icmp.pcap.gz has this packet too. TCP SYN packet with multicast src address.

Comment (tbursztyka) [reply subsys/net/ip/tcp.c]: Ok, but that's not what #11329 is talking about, there it's only about arp table that should not be updated etc... 

If that syn/mcast is a bug, it's a different one. So do not set Fixes #11329

**Review by tbursztyka — APPROVED** [review]

**Comment by rveerama1** [comment]
@nashif ping

**Comment by pfalcon** [comment]
Note that I continue to insist that proper logging is added to the drop cases. We have enough cases of silent dropping already, and when something is logged, it's confusing like hell. So, existing cases should be fixed, like e.g. done in https://github.com/zephyrproject-rtos/zephyr/pull/11662, not more silent-behind-the-user's-back cases added.

**Comment by tbursztyka** [comment]
@rveerama1 It kind of make sense so dbg log drop reason actually, can you add such message? (see my comment in PR #11662 to "format" the message )

**Comment by rveerama1** [comment]
Added logging messages.

File: subsys/net/l2/ethernet/arp.c
Comment (pfalcon) [inline subsys/net/l2/ethernet/arp.c]: Thanks for adding logging. But per @tbursztyka's proposal, it should be:

~~~
NET_DBG("DROP: eth addr is bcast, src addr is mcast");
~~~

I think it makes sense to follow that proposal.

Comment (rveerama1) [reply subsys/net/l2/ethernet/arp.c]: [00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): DROP: eth addr is bcast, src addr is mcast

Or

[00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): Drop: eth addr is bcast, src addr is mcast

Or

[00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): Drop, eth addr is bcast, src addr is mcast

Comment (rveerama1) [reply subsys/net/l2/ethernet/arp.c]: decide which one make sense?

Comment (tbursztyka) [reply subsys/net/l2/ethernet/arp.c]: first one

Comment (pfalcon) [reply subsys/net/l2/ethernet/arp.c]: Oops, sorry, forgot to link to his post: https://github.com/zephyrproject-rtos/zephyr/pull/11662#discussion_r236543329 . There, he clearly shows `DROP: ` syntax. But can wait for extra confirmation, ping @tbursztyka .

Comment (rveerama1) [reply subsys/net/l2/ethernet/arp.c]: ok

**Comment by rveerama1** [comment]
Updated.

**Comment by rveerama1** [comment]
Updated.

**Review by pfalcon — APPROVED** [review]
Thanks for updates!

**Comment by tbursztyka** [comment]
@pfalcon I am not able to merge anything so let's wait for @nashif or whoever has merge rights.

**Comment by pfalcon** [comment]
@tbursztyka:

> @pfalcon I am not able to merge anything so let's wait for @nashif or whoever has merge rights.

Really? I thought you used to have. Anyway, I didn't ask to merge, just checked if you fully understand the changes in https://github.com/zephyrproject-rtos/zephyr/commit/6c872cbd75e1af78f9b8f5144f25f73fd805e6fd of this patch (because I don't from just looking at it).

**Comment by tbursztyka** [comment]
@pfalcon I never had merge rights on zephry. About commit 6c872cb, I do understand the change. Actually, while doing my PR #10547 I re-factored ethernet_send (current one is quite a mess), and I did fix that issue by "chance" (Since the code is clearer, is was obvious what to set as ptype there). So current fix is right, but hopefully will switch to a better ethernet_send soon.

**Comment by pfalcon** [comment]
@carlescufi, Maybe you could merge this?

**Pull request merged**
